src/ows/src/main/java/org/geoserver/ows/Dispatcher.java

```diff
@@ -1159,6 +1197,20 @@ Service findService(String id, String ver, String namespace) throws ServiceExcep
 
         Service sBean = null;
 
+        if ((version == null) && (acceptsVersions != null)) {
+            // version negotiation
+            // 06-121r3: The server, upon receiving a GetCapabilities request, shall scan through
+            // this list and find the first version number that it supports.
```

Comment:
Agree. I would like to see a test of this behavior though. The tests seem to check parsing occurs correctly, but I don't see a test checking "the first version the server supports" has been chosen. I guess it would be easy-ish to write in the WFS module, which sports 3 different WFS versions in a single module.

Guess what, I've found a VersionNegotiationTest in WFS, so I'm guessing the WFS machinery also has code to do negotiation... oh, it's actually in a RequestUtils method it seems (so possibly some duplication with the work happening here):
https://github.com/geoserver/geoserver/blob/de184b2f9ed9a844b5291dd163e772132a07689b/src/wfs/src/main/java/org/geoserver/wfs/GetCapabilities.java#L97

Reply:
Trying to remove this with david showed a number different things:

* Some version negotiation happens in KVP parsing; producing OWS11Version data structures which do not work correctly when talking to WFS 1.0
* Reducing everything to a single acceptsVersion method; and then updating the KVP works ... but then a whole bunch of test cases across the codebase fail - where their are testing the accepts version functionality that has been implemented separately in the individual modules. While this is expected as the code is obsolete / removed
* When WFS 0.8 is provided OWS1.0 version negotiation expects 1.0 (closest), while OWS1.1 expects 2.0 (highest) so some changes were needed.

I love the idea of Dispatcher handling acceptsVersions; but I think each protocol that has its own logic would need to migrate over to trusting dispatcher to do the right thing.

I am not clear how (or if) fixing the KVP Parser version handling should be done; WFS 1.0 was producing a strongly typed OWS11 Version object rather than a string.

Reply:
I expect that WFS code was used to process the ``http://localhost/geoserver/wfs?`` URLs (when ``SERVICE=WFS`` is not included). In any case dispatcher change is a good approach; but the other modules doing their own accceptsVersions negotiation may need to be migrated over time; and have their test cases adjusted to match.